prov/efa/test/efa_unit_test_hmem.c

```diff
@@ -49,9 +49,109 @@ void test_efa_hmem_info_update_neuron(struct efa_resource **state)
         assert_false(efa_domain->hmem_info[FI_HMEM_NEURON].initialized);
         assert_false(efa_domain->hmem_info[FI_HMEM_NEURON].p2p_supported_by_device);
 }
+
+/**
+ * @brief Verify when p2p is disabled, we don't check p2p support with neuron_alloc.
+ * Just leave p2p_supported_by_device to false.
+ *
+ * @param[in]	state		struct efa_resource that is managed by the framework
+ */
+void test_efa_hmem_info_disable_p2p_neuron(struct efa_resource **state)
+{
+        int ret;
+        struct efa_resource *resource = *state;
+        struct efa_domain *efa_domain;
+        uint32_t efa_device_caps_orig;
+        bool neuron_initialized_orig;
+
+	resource->hints = efa_unit_test_alloc_hints(FI_EP_RDM);
+        assert_non_null(resource->hints);
+
+        ret = fi_getinfo(FI_VERSION(1, 14), NULL, NULL, 0ULL, resource->hints, &resource->info);
+        assert_int_equal(ret, 0);
+
+        ret = fi_fabric(resource->info->fabric_attr, &resource->fabric, NULL);
+        assert_int_equal(ret, 0);
+
+        neuron_initialized_orig = hmem_ops[FI_HMEM_NEURON].initialized;
+        hmem_ops[FI_HMEM_NEURON].initialized = true;
+        efa_device_caps_orig = g_device_list[0].device_caps;
+        g_device_list[0].device_caps |= EFADV_DEVICE_ATTR_CAPS_RDMA_READ;
+        /* neuron_alloc should not be called when p2p is disabled */
+        g_efa_unit_test_mocks.neuron_alloc = &efa_mock_neuron_alloc_return_mock;
+
+        ofi_hmem_disable_p2p = 1;
+        ret = fi_domain(resource->fabric, resource->info, &resource->domain, NULL);
+
+        /* recover the modified global variables before doing check */
+        ofi_hmem_disable_p2p = 0;
+        g_device_list[0].device_caps = efa_device_caps_orig;
+        hmem_ops[FI_HMEM_NEURON].initialized = neuron_initialized_orig;
+
+        assert_int_equal(ret, 0);
+        efa_domain = container_of(resource->domain, struct efa_domain,
+                                  util_domain.domain_fid.fid);
+        assert_true(efa_domain->hmem_info[FI_HMEM_NEURON].p2p_disabled_by_user);
+        assert_true(efa_domain->hmem_info[FI_HMEM_NEURON].initialized);
+        assert_false(efa_domain->hmem_info[FI_HMEM_NEURON].p2p_supported_by_device);
+}
 #else
 void test_efa_hmem_info_update_neuron()
 {
         skip();
 }
+
+void test_efa_hmem_info_disable_p2p_neuron()
+{
+        skip();
+}
 #endif /* HAVE_NEURON */
+
+#if HAVE_CUDA
```

Comment:
A topic for later - do we really need `HAVE_CUDA/HAVE_NEURON`? If we are calling `ofi_cudaMalloc/ofi_neuronMalloc`, then we don't need to check CUDA/Neuron availability again.